platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/RuntimeWorkspaceRegistry.java

```diff
@@ -334,54 +306,123 @@ public RuntimeWorkspaceImpl get(String workspaceId) throws NotFoundException {
     /**
      * Starts all environment machines, starting from dev machine.
      */
-    List<MachineImpl> startEnvironment(Environment environment, String workspaceId, boolean recover) throws BadRequestException,
-                                                                                                            ServerException,
-                                                                                                            NotFoundException,
-                                                                                                            ConflictException {
-        /* todo replace with environments management
-           for now we consider environment is:
-           - recipe with type "docker"
-           - recipe script will be ignored
-           - list of docker machines configs
-           - one of machines in this list is dev machine
-        */
-        //FIXME "docker"
-        String envRecipeType = environment.getRecipe() == null ? "docker" : environment.getRecipe().getType();
-        if (!"docker".equals(envRecipeType)) {
-            throw new BadRequestException("Invalid environment recipe type " + envRecipeType);
+    private void startEnvironment(Environment environment, String workspaceId, boolean recover) throws BadRequestException,
+                                                                                                       ServerException,
+                                                                                                       NotFoundException,
+                                                                                                       ConflictException {
+        final List<? extends MachineConfig> machineConfigs = new ArrayList<>(environment.getMachineConfigs());
+        final MachineConfig devCfg = findDev(machineConfigs);
+
+        // Starting workspace dev machine
+        final MachineImpl devMachine;
+        try {
+            devMachine = createMachine(devCfg, workspaceId, environment.getName(), recover);
+        } catch (MachineException | BadRequestException | NotFoundException | SnapshotException | ConflictException ex) {
+            doRemoveWorkspace(workspaceId);
+            throw ex;
         }
 
-        final List<MachineImpl> machines = new ArrayList<>();
+        // Updating workspace status and machines list
+        lock.writeLock().lock();
+        try {
+            addRunningMachine(devMachine);
+            final RuntimeWorkspaceImpl workspace = idToWorkspaces.get(workspaceId);
+            if (workspace != null) {
```

Comment:
Nice remark.

Workspace may be `null` if and only if `@PreDestroy` was performed and workspace was removed from in-memory storage, so it is the same to `if (isStopping)`, i missed that machine won't be stopped by `@PreDestroy` method as it stops all machines that were added to `RuntimeWorkspace` instance, but in this case `devMachine` hasn't been added yet, so it won't be destroyed. I am going to fix it, thank you.